### DISCUSSION TITLE
Windhawk.dll in every process?

### DISCUSSION THREAD

**Discussion by ErmineMD** [ramensoftware/windhawk#21] [General]
Windhawk.dll is in every process. Is it really necessary? I don’t use any global mods that interact with all apps. Only Large Taskbar Icons and Taskbar Volume Control.
Is it possible not to inject it if unnecessary? Or is it possible to have a switch that turns it off completely?

Update.
Oh. I‘ve just read your answer to another question about it.

**Comment by m417z** [comment]
The reason Windhawk injects its module into all processes is to be able to customize programs at a very early stage. It might be important, or not so much, depending on the mod.

For example, in order to be able to customize Notepad, launched by Explorer, at a very early stage, I need to be injected into Explorer to be able to hijack the process creation.

For the technical details about how it all works, you can check out the following blog post: [Implementing Global Injection and Hooking in Windows](https://m417z.com/Implementing-Global-Injection-and-Hooking-in-Windows/).

Having said that, I understand that some users might prefer limiting Windhawk to specific processes, and some users must exclude processes which are incompatible. I'm working on adding options to exclude processes from Windhawk for the next version.

**Reply by shoaibshakeel381** [reply]
Anyway to only inject into processes for which mods are installed?

**Reply by m417z** [reply]
Not right now, but I plan to add an option for it in the next version of Windhawk.

**Comment by m417z** [comment]
With Windhawk v1.0, it's now possible to configure a list of excluded processes.

**Reply by brgr-m-r** [reply]
Perhaps a basic question since I'm new to this:

How can we identify all relevant processes for which certain mods are installed (e.g. via Process Hacker)?
And configure Windhawk to inject only into those?

For example, if only 'Taskbar Thumbnail Reorder' mod is required, how should I configure the Windhawk process inclusion/exclusion lists? I assume just adding explorer.exe to the inclusion list is not enough for that.

It would be great to minimize any side-effects. FYI: recently I got a major nightmarish issue with explorer.exe recurrently crashing (with unclear dump errors) with both 7+ Taskbar Tweaker / ExplorerPatcher (required combo, initially worked fine) and then Windhawk (0.9.2). Win 11 system reinstall with in-place upgrade (keeping apps) didn't help. Seems resolved only by completely uninstalling Windhawk and completely cleaning all types of temporary files. The system was clean and stable before all that started.

Many thanks.

**Reply by m417z** [reply]
> How can we identify all relevant processes for which certain mods are installed (e.g. via Process Hacker)?

The easiest way is just [to look at the mod target processes](https://i.imgur.com/X6HVOXI.png). For Taskbar Thumbnail Reorder, there's one target process, explorer.exe.

If you want to see which mods are injected into which processes at any given time, you can see that in the Loaded mods dialog. It can be opened from the tray icon right click menu.

> And configure Windhawk to inject only into those?

You can configure it [in the advanced settings](https://i.imgur.com/ktPDUNn.png). You can set `*` - all processes - in the exclusion list, and `explorer.exe` in the inclusion list. Note that this will cause mods to be injected into Explorer with a slight delay, as I explained in the comment above. For Taskbar Thumbnail Reorder that isn't that important, but for other mods it might be.

> FYI: recently I got a major nightmarish issue with explorer.exe recurrently crashing (with unclear dump errors) with both 7+ Taskbar Tweaker / ExplorerPatcher (required combo, initially worked fine) and then Windhawk (0.9.2).

Incompatibilities like this happen sometimes, but disabling all relevant Windhawk mods and exiting 7+ Taskbar Tweaker should be enough, there's no need to uninstall the programs or delete temporary files. One of the advantages of both Windhawk and 7+ Taskbar Tweaker is that they don't modify system files or registry entries, so as long as they don't run, no changes to the system are made.

**Reply by albino1** [reply]
I hit this today where a program wouldn't exit on the command line even when explicitly specifying exit with `/x`.  Took a while to track this change in behavior down to Windhawk, I just had to figure out what I had installed recently on both my laptop (Win 10) and my PC (Win 11) and start uninstalling and reinstalling things.

I added the processes as excluded in Windhawk and voila everything works!  What a great piece of software, thanks for maintaining it.
